### PR TITLE
Archivist: Mirror --recent mode

### PR DESCRIPTION
**Pull request by graydon**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`

### Thoroughness

* [x] This PR adds tests for the most critical parts of the new functionality or fixes.
* [x] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [x] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [x] (patch) I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### Summary

Add a `--recent` mode for `mirror`. This saves the user from guessing the appropriate `N` for `--last N`, or a range for `--low` / `--high`. Simply call `stellar-archivist mirror --recent src dst` and it will copy the range from dst's most recent HAS to src's. Running this in a loop should be sufficient to incrementally mirror archives.

### Goal and scope

Make it easier to incrementally mirror an archive.

### Summary of changes

One new flag and bugfix / logging changes.

### Known limitations & issues

It doesn't (and no mode of archivist currently does) do an exact delta between the archives. If something is corrupt or missing in the dst -- but not identified by just "copying the range difference" -- this doesn't find or fix it. A general, fast cross-archive repair mechanism doesn't exist yet.

### What shouldn't be reviewed

N/A

### CODE REVIEW

File: tools/stellar-archivist/main.go [hunk 1]
```diff
@@ -43,19 +43,32 @@ type Options struct {
...
+			if e1 == nil && e2 == nil {
+				low := dstState.CurrentLedger
+				high := srcState.CurrentLedger
+				opts.CommandOpts.Range =
+					historyarchive.MakeRange(low, high)
+				return
+			}
+		} else if opts.Last != -1 {
+			state, e := srcArch.GetRootHAS()
+			if e != nil {
```
Comment (ire-and-curses) [inline tools/stellar-archivist/main.go]: should this be `if e == nil`? I guess the fact I'm not following the logic here indicates a comment for each if block would be helpful.

Comment (graydon) [reply tools/stellar-archivist/main.go]: Ah good catch! fixing and commenting.

**Comment by graydon** [comment]
Re: testing, yeah, I forgot there is some rudimentary testing support for archivist, I should flesh it out a bit to support this PR at least. The other one (#1637 -- S3 fixes) is I think not within the realm of unit-testability.

**Comment by ire-and-curses** [comment]
> Re: testing, yeah, I forgot there is some rudimentary testing support for archivist, I should flesh it out a bit to support this PR at least. The other one (#1637 -- S3 fixes) is I think not within the realm of unit-testability.

Ok - add me as a reviewer again when you've added them

**Comment by graydon** [comment]
@ire-and-curses added tests and rebased, should be ready to land.

**Pull request merged**